### PR TITLE
add support for extracting tags from metric key by javascript RegExp

### PR DESCRIPTION
**Pull request by oryband**

### CODE REVIEW

File: lib/datadog.js [hunk 1]
```diff
@@ -225,6 +290,17 @@ exports.init = function datadog_init(startup_time, config, events, log) {
       datadogApiHost = 'https://app.datadoghq.com';
    }
 
+   // Read metric regexps string into RegExp form.
+   if (config.datadogMetricTagsByRegexp !== undefined &&
+       config.datadogMetricTagsByRegexp.constructor === Object || 
+       Objects.keys(config.datadogMetricTagsByRegexp).length >= 1) {
+
+       for (re in config.datadogMetricTagsByRegexp) {
```
Comment (liorama) [inline lib/datadog.js]: `let re`

File: lib/datadog.js [hunk 2]
```diff
@@ -28,6 +50,7 @@ var datadogApiKey;
 var datadogStats = {};
 var datadogPrefix;
 var datadogTags;
+var datadogMetricTagsByRegexp;
```
Comment (liorama) [inline lib/datadog.js]: I know it isn't part of this pr, but can you split the variable defenestrations at lines 40-42 into different statements?

File: lib/datadog.js [hunk 3]
```diff
@@ -53,18 +76,13 @@ Datadog.prototype._post = function(controller, message) {
         logger.log('Warning! You are about to send unencrypted metrics.');
     }
     client.pending_requests += 1;
-    request.post({
+    console.log({
```
Comment (liorama) [inline lib/datadog.js]: Is this meant to stay here? maybe wrap it in a function which checks `env == debug` and then logs?

File: lib/datadog.js [hunk 4]
```diff
@@ -90,32 +108,36 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
    var payload = [];
    var value;
 
-   var key;
+   var key,
+       metricName,
+       metricTags;
```
Comment (liorama) [inline lib/datadog.js]: split up into different statments

File: lib/datadog.js [hunk 5]
```diff
@@ -90,32 +108,36 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
 
    // Send counters
    for (key in counters) {
```
Comment (liorama) [inline lib/datadog.js]: `let key` or `var key` otherwise it'll be global (true for all occurences). 
Also, If this is't a sterile environment, ie an environment provided by datadog and
they could've change the prototype for `Object` then it's worth doing:
`Object.keys(counters).forEach(function(key){...})` so you won't get bogus keys they've added.
You can try it like you've done and if weird shit happens remember it might be this.

File: lib/datadog.js [hunk 6]
```diff
@@ -90,32 +108,36 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
       value = counters[key];
       var valuePerSecond = value / (flushInterval / 1000); // calculate 'per second' rate
+      metricName, metricTags = get_metric(key);
```
Comment (liorama) [inline lib/datadog.js]: Did this work!? It shouldn't. This is a new concept in JS but I'm pretty sure you need to do it like this:
`[metricName, metricTags] = get_metric(key);`

File: lib/datadog.js [hunk 7]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+         metricName, metricTags = get_metric(key + '.count');
          payload.push({
-            metric: get_prefix(key + '.count'),
+            metric: metricKey,
             points: [[ts, count]],
             type: 'gauge',
             host: host,
-            tags: datadogTags
+            tags: datadogTags.concat(metricTags)
          });
```
Comment (liorama) [inline lib/datadog.js]: Instead of having the same code duplicated many times just have a function called on an array of suffixes:
```
var metrixSuffixList = ['mean', 'upper', .upper_' + pctThreshold...]
metrixSuffixList.forEach(function(suffix){
[metricName, metricKey] = get_metric(key + "." + suffix)
payload.push({
             metric: get_prefix(key + '.count'),
             metric: metricKey,
             points: [[ts, count]],
             type: 'gauge',
             host: host,
             tags: datadogTags
             tags: datadogTags.concat(metricTags)
          });
})
```

File: lib/datadog.js [hunk 8]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
       }
    }
 
    post_stats(payload);
 };
 
+// Returns the metric name according to configuration.
+var get_metric = function datadog_get_metric(key) {
```
Comment (liorama) [inline lib/datadog.js]: Move all function expressions to the top of the scope.
Also, why do use give the function a name and put it into a var too? It isn't necessarily bad but you need to know why you're doing it.

File: lib/datadog.js [hunk 9]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+    var metricKey, metricTags;
```
Comment (liorama) [inline lib/datadog.js]: split into different statments

File: lib/datadog.js [hunk 10]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+    metricKey, metricTags = get_metric_tags_by_regexp(key);
```
Comment (liorama) [inline lib/datadog.js]: `[]` around array destructuring (the return values)

File: lib/datadog.js [hunk 11]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+    metricKey = get_prefix(metricKey);
+    return metricKey, metricTags;
```
Comment (liorama) [inline lib/datadog.js]: here too:
`return [metricKey, metricTags];'

File: lib/datadog.js [hunk 12]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+
+// Tries to match key with all metric RegExps found in configuration.
+// When a RegExp is matched, its captured groups are truncated from the metric
+// name and are pushed into the metric's tag array.
+//
+// Returns the mutated key and its tag array.
+// If no matchis found, the key is returned as-is with an empty tag array.
+var get_metric_tags_by_regexp = function datadog_get_metric_tags_by_regexp(key) {
+    var match;
+    for (re in datadogMetricTagsByRegexp) {
```
Comment (liorama) [inline lib/datadog.js]: `let re`

File: lib/datadog.js [hunk 13]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+        if (datadogMetricTagsByRegexp.hasOwnProperty(re)) {
+            match = re.exec(key);
+            if (match !== null && match.length - 1 > 0) {
```
Comment (liorama) [inline lib/datadog.js]: just do:
`if (match && match.length > 1) {`

File: lib/datadog.js [hunk 14]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+                var mutatedKey = key,
```
Comment (liorama) [inline lib/datadog.js]: `var` statements should be placed the top of the scope. Move those bitches up and split into different statements.

File: lib/datadog.js [hunk 15]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+                    metricTags = [],
+                    tagNameArray = datadogMetricTagsByRegexp[re];
+
+                for (var i = 1; i < match.length; i++) {
```
Comment (liorama) [inline lib/datadog.js]: Nice, but we are not in the 70s and this ain't c++:
```match.forEach(function(tagNames){
    metricTags.push(tagNameArray[i-1] + ':' + tagNames.join(':'));
    mutatedKey = mutatedKey.replace(new RegExp(tagNames+"\\.?", "")) // you don't have to pass ""
})```

File: lib/datadog.js [hunk 16]
```diff
@@ -146,51 +168,93 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+                    metricTags.push(tagNameArray[i-1] + ':' + match[i].join(':'));
+                    mutatedKey = mutatedKey.replace(new RegExp(match[i]+"\\.?", ""))
+                }
+            }
+
+            return mutatedKey, metricTags
+        }
+    }
+
+    return key, []
```
Comment (liorama) [inline lib/datadog.js]: here too

File: lib/datadog.js [hunk 17]
```diff
@@ -225,6 +290,17 @@ exports.init = function datadog_init(startup_time, config, events, log) {
       datadogApiHost = 'https://app.datadoghq.com';
    }
 
+   // Read metric regexps string into RegExp form.
+   if (config.datadogMetricTagsByRegexp !== undefined &&
+       config.datadogMetricTagsByRegexp.constructor === Object || 
+       Objects.keys(config.datadogMetricTagsByRegexp).length >= 1) {
```
Comment (liorama) [inline lib/datadog.js]:
```
if (config.datadogMetricTagsByRegexp &&
 Objects.keys(config.datadogMetricTagsByRegexp).length > 0)
```
also, Why are you checking for the constructor? is there a chance it won't be an object?

File: lib/datadog.js (hunk 1 above)
Comment (oryband) [inline lib/datadog.js]: is this node 4

File: lib/datadog.js (hunk 2 above)
Comment (oryband) [inline lib/datadog.js]: like you said - this'll have to go into another pr since i plan to pr this upstream and try to keep changes to a minimum

File: lib/datadog.js (hunk 3 above)
Comment (oryband) [inline lib/datadog.js]: will be removed, notice there's a "debug" commit

File: lib/datadog.js (hunk 5 above)
Comment (oryband) [inline lib/datadog.js]: note that the lines which aren't included in the diff have been there to begin with - i didn't write this

File: lib/datadog.js (hunk 7 above)
Comment (oryband) [inline lib/datadog.js]: yeah i thought of doing this as well - but this isn't the pr's topic

File: lib/datadog.js (hunk 8 above)
Comment (oryband) [inline lib/datadog.js]: copied the same behaviour from the file, why is this bad?

File: lib/datadog.js (hunk 15 above)
Comment (oryband) [inline lib/datadog.js]: you should comment below the code, not above it. i can't see what you commented about now

File: lib/datadog.js (hunk 17 above)
Comment (oryband) [inline lib/datadog.js]: yes, i can't predict what people will put in the configuration file

File: lib/datadog.js (hunk 8 above)
Comment (liorama) [inline lib/datadog.js]: It'll be hoisted up by the js parser, most times it won't be a problem but it can cause some weird behaviour. It's good practice.
http://www.adequatelygood.com/JavaScript-Scoping-and-Hoisting.html

**Comment by oryband** [comment]
this pr was only opened for inner-rounds comments, not for merging. opened DataDog/statsd-datadog-backend#14 for merging to master

File: lib/datadog.js (hunk 7 above)
Comment (liorama) [inline lib/datadog.js]: Do it anyway it's a small enough but yet helpful refactoring

File: lib/datadog.js [hunk 18]
```diff
@@ -146,51 +182,137 @@ var flush_stats = function datadog_post_stats(ts, metrics) {
...
+// If no match is found, the key is returned as-is with an empty tag array.
+var get_metric_tags_by_regexp = function datadog_get_metric_tags_by_regexp(key) {
+    var match;
+
+    for (var i = 0; i < datadogMetricTagsByRegExp.length; i++) {
+        var regExpTags = datadogMetricTagsByRegExp[i];
+        var re = regExpTags['regExp'];
+        // RegExp matches start from index i=1 instead of i=0,
+        // so the tag names that will be matched are offset accordingly.
+        var tagNames = [ null ].concat(regExpTags['tagNames']);
```
Comment (liorama) [inline lib/datadog.js]: `var` statements should always be place at the top of the scope

File: lib/datadog.js [hunk 19]
```diff
@@ -225,6 +348,23 @@ exports.init = function datadog_init(startup_time, config, events, log) {
       datadogApiHost = 'https://app.datadoghq.com';
    }
 
+   // Read metric regexps string into RegExp form.
+   if (config.datadogMetricTagsByRegExp &&
+       config.datadogMetricTagsByRegExp.constructor === Object &&
```
Comment (liorama) [inline lib/datadog.js]: `config.datadogMetricTagsByRegExp.constructor === Object` why?

File: lib/datadog.js [hunk 20]
```diff
@@ -225,6 +348,23 @@ exports.init = function datadog_init(startup_time, config, events, log) {
...
+       Object.keys(config.datadogMetricTagsByRegExp).length > 0) {
+
+       for (var re in config.datadogMetricTagsByRegExp) {
```
Comment (liorama) [inline lib/datadog.js]: move `var re` up

File: lib/datadog.js [hunk 21]
```diff
@@ -225,6 +348,23 @@ exports.init = function datadog_init(startup_time, config, events, log) {
...
+           if (config.datadogMetricTagsByRegExp.hasOwnProperty(re)) {
+               datadogMetricTagsByRegExp.push(
+                   {
```
Comment (liorama) [inline lib/datadog.js]: `datadogMetricTagsByRegExp.push({`

File: lib/datadog.js [hunk 22]
```diff
@@ -225,6 +348,23 @@ exports.init = function datadog_init(startup_time, config, events, log) {
...
+                       'regExp': new RegExp(re),
+                       'tagNames': config.datadogMetricTagsByRegExp[re]
+                   }
+               );
```
Comment (liorama) [inline lib/datadog.js]: })
Configure vim to do this for you.

File: lib/datadog.js [hunk 23]
```diff
@@ -216,6 +338,7 @@ exports.init = function datadog_init(startup_time, config, events, log) {
    datadogApiHost = config.datadogApiHost;
    datadogPrefix = config.datadogPrefix;
    datadogTags = config.datadogTags;
+   datadogMetricTagsByRegExp = [];
```
Comment (liorama) [inline lib/datadog.js]: Is this a global variable?

File: lib/datadog.js (hunk 7 above)
Comment (oryband) [inline lib/datadog.js]: it's actually done in https://github.com/rounds/statsd-datadog-backend/commit/3bcdf08f7f86c736583d4ad0e1d0315a55b3eb6f

File: lib/datadog.js (hunk 19 above)
Comment (oryband) [inline lib/datadog.js]: to make sure the correct configuration type was given. i'm expecting an object and not an array for example

File: lib/datadog.js (hunk 22 above)
Comment (oryband) [inline lib/datadog.js]: actually i've done this manually, but ok
